lib/fauna/connection.rb

```diff
@@ -95,58 +132,55 @@ def query_string_for_logging(query)
       end.join('&')
     end
 
-    def inflate(response)
-      if %w(gzip deflate).include?(response.headers['Content-Encoding'])
-        Zlib::GzipReader.new(StringIO.new(response.body.to_s), :external_encoding => Encoding::UTF_8).read
+    def execute(action, path, query = nil, data = nil) # rubocop:disable Metrics/MethodLength
+      if @logger.empty?
+        response = execute_without_logging(action, path, query, data)
       else
-        response.body.to_s
-      end
-    end
-
-    def execute(action, ref, data = nil, query = nil) # rubocop:disable Metrics/CyclomaticComplexity, Metrics/MethodLength
-      if @logger
-        log(0) { "Fauna #{action.to_s.upcase} /#{ref}#{query_string_for_logging(query)}" }
+        log(0) { "Fauna #{action.to_s.upcase} /#{path}#{query_string_for_logging(query)}" }
         log(2) { "Credentials: #{@credentials}" }
         log(2) { "Request JSON: #{JSON.pretty_generate(data)}" } if data
 
-        t0, r0 = Process.times, Time.now
-        response = execute_without_logging(action, ref, data, query)
-        t1, r1 = Process.times, Time.now
-        body = inflate(response)
+        t0 = Time.now
+        response = execute_without_logging(action, path, query, data)
+        t1 = Time.now
 
-        real = r1.to_f - r0.to_f
-        cpu = (t1.utime - t0.utime) + (t1.stime - t0.stime) + (t1.cutime - t0.cutime) + (t1.cstime - t0.cstime)
-        log(2) { ["Response headers: #{JSON.pretty_generate(response.headers)}", "Response JSON: #{body}"] }
-        log(2) { "Response (#{response.status}): API processing #{response.headers['X-HTTP-Request-Processing-Time']}ms, network latency #{((real - cpu) * 1000).to_i}ms, local processing #{(cpu * 1000).to_i}ms" }
-      else
-        response = execute_without_logging(action, ref, data, query)
-        body = inflate(response)
+        network_latency = t1.to_f - t0.to_f
+        log(2) { ["Response headers: #{JSON.pretty_generate(response.headers)}", "Response JSON: #{response.body}"] }
+        log(2) { "Response (#{response.status}): API processing #{response.headers['X-HTTP-Request-Processing-Time']}ms, network latency #{(network_latency * 1000).to_i}ms" }
       end
 
-      case response.status
-      when 200..299
-        [response.headers, body]
-      when 400
-        fail BadRequest.new(JSON.parse(body))
-      when 401
-        fail Unauthorized.new(JSON.parse(body))
-      when 403
-        fail PermissionDenied.new(JSON.parse(body))
-      when 404
-        fail NotFound.new(JSON.parse(body))
-      when 405
-        fail MethodNotAllowed.new(JSON.parse(body))
-      else
-        fail NetworkError, body
+      response
+    end
+
+    def execute_without_logging(action, path, query, data)
+      @connection.send(action) do |req|
+        req.params = query.delete_if { |_, v| v.nil? || v.empty? } unless query.nil?
```

Comment:
Url query params by their nature have to be strings. I can skip removing non-nil values though.